rmw_zenoh_cpp/src/zenohd/main.cpp

```diff
@@ -39,135 +35,25 @@
-    }
-#else
-    // get the console in raw mode
-    if (tcgetattr(0, &cooked_) < 0) {
-      throw std::runtime_error("Failed to get old console mode");
-    }
-    struct termios raw;
-    memcpy(&raw, &cooked_, sizeof(struct termios));
-    raw.c_lflag &= ~(ICANON | ECHO);
-    // Setting a new line, then end of file
-    raw.c_cc[VEOL] = 1;
-    raw.c_cc[VEOF] = 2;
-    raw.c_cc[VTIME] = 1;
-    raw.c_cc[VMIN] = 0;
-    if (tcsetattr(0, TCSANOW, &raw) < 0) {
-      throw std::runtime_error("Failed to set new console mode");
-    }
-#endif
-  }
-
-  char readOne()
-  {
-    char c = 0;
-
-#ifdef _WIN32
-    INPUT_RECORD record;
-    DWORD num_read;
-    switch (WaitForSingleObject(hstdin_, 100)) {
-      case WAIT_OBJECT_0:
-        if (!ReadConsoleInput(hstdin_, &record, 1, &num_read)) {
-          throw std::runtime_error("Read failed");
-        }
-
-        if (record.EventType != KEY_EVENT || !record.Event.KeyEvent.bKeyDown) {
-          break;
-        }
-
-        if (record.Event.KeyEvent.wVirtualKeyCode == VK_LEFT) {
-          c = KEYCODE_LEFT;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == VK_UP) {
-          c = KEYCODE_UP;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == VK_RIGHT) {
-          c = KEYCODE_RIGHT;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == VK_DOWN) {
-          c = KEYCODE_DOWN;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == 0x42) {
-          c = KEYCODE_B;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == 0x43) {
-          c = KEYCODE_C;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == 0x44) {
-          c = KEYCODE_D;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == 0x45) {
-          c = KEYCODE_E;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == 0x46) {
-          c = KEYCODE_F;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == 0x47) {
-          c = KEYCODE_G;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == 0x51) {
-          c = KEYCODE_Q;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == 0x52) {
-          c = KEYCODE_R;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == 0x54) {
-          c = KEYCODE_T;
-        } else if (record.Event.KeyEvent.wVirtualKeyCode == 0x56) {
-          c = KEYCODE_V;
-        }
-        break;
-
-      case WAIT_TIMEOUT:
-        break;
-    }
-
-#else
-    int rc = read(0, &c, 1);
-    if (rc < 0) {
-      throw std::runtime_error("read failed");
-    }
-#endif
-
-    return c;
-  }
-
-  ~KeyboardReader()
-  {
-#ifdef _WIN32
-    SetConsoleMode(hstdin_, old_mode_);
-#else
-    tcsetattr(0, TCSANOW, &cooked_);
-#endif
-  }
-
-private:
-#ifdef _WIN32
-  HANDLE hstdin_;
-  DWORD old_mode_;
-#else
-  struct termios cooked_;
-#endif
-};
+static std::mutex run_mutex;
+static std::condition_variable run_cv;
 
 #ifdef _WIN32
 BOOL WINAPI quit(DWORD ctrl_type)
 {
   (void)ctrl_type;
+  std::scoped_lock lock(run_mutex);
```

Comment:
So we can't (safely) take a mutex inside of a signal handler, because it is possible for the mutex to block.

Luckily, we don't need to take the mutex in order to run `cv.notify_one()`.  Indeed, the documentation for it says that it is a [pessimization](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one) (see the Notes).  Thus I think this should be removed:

```suggestion
```